### PR TITLE
tfidf function and test

### LINKED ISSUE #27: Code and Test for `tfidf_vectorizer`

### PR DESCRIPTION
**Pull request by MoNorouzi23**
updating tfidf function and tests

### CODE REVIEW

File: src/text_processing_util_mds24/text_processing_util_mds24.py [hunk 1]
```diff
@@ -1,5 +1,7 @@
 import string
 import numpy as np
+from collections import Counter
+from sklearn.feature_extraction.text import TfidfVectorizer
```
Comment (jy1909) [inline src/text_processing_util_mds24/text_processing_util_mds24.py]: Remove this line (line 4) since `TfidfVectorizer` is not used

**Review by jy1909 — COMMENTED** [review]
I just looked at the tests.

1. There are some edge cases at the top of the test file (lines 6-11). Let's test these edge cases as well.
2. I see we are just checking the length of the array here. We can use `numpy.testing.assert_array_equal` and `numpy.testing.assert_array_almost_equal` to check for the content of the array. Links: https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html and https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_almost_equal.html

File: src/text_processing_util_mds24/text_processing_util_mds24.py [hunk 2]
```diff
@@ -1,5 +1,10 @@
 import string
 import numpy as np
+from collections import Counter
+<<<<<<< HEAD
+
+=======
+>>>>>>> origin/27-code-and-test-for-tfidf_vectorizer
```
Comment (jy1909) [inline src/text_processing_util_mds24/text_processing_util_mds24.py]: Remove
```suggestion

```

File: src/text_processing_util_mds24/text_processing_util_mds24.py [hunk 3]
```diff
@@ -126,3 +156,5 @@ def tokenizer_padding(docs: list[str]) -> np.ndarray:
     [[1, 2, 3], [2, 3, 4]]
     """
     pass
+docs_1= ["apple orange banana", "apple banana banana"]
+tfidf_vectorizer(docs_1)
```
Comment (jy1909) [inline src/text_processing_util_mds24/text_processing_util_mds24.py]: Remove
```suggestion

```

File: tests/test_text_processing_util_mds24.py [hunk 4]
```diff
@@ -80,8 +83,84 @@ def test_list_num_punctuation():
...
+
+def test_tfidf_vectorizer_one_document():
+    docs = ["python is a programming language"]
+    tfidf_matrix, feature_names = tfidf_vectorizer(docs)
+
+    expected_matrix = np.array([
+        [-0.13862944, -0.13862944, -0.13862944, -0.13862944, -0.13862944]
+    ])
+    assert_array_almost_equal(tfidf_matrix, expected_matrix[:, :5], decimal=6)
+    assert_array_equal(feature_names, np.array(['a', 'is', 'language', 'programming', 'python']))
```
Comment (jy1909) [inline tests/test_text_processing_util_mds24.py]: Add `np.testing.` because the extra import was removed.
```suggestion
    np.testing.assert_array_almost_equal(tfidf_matrix, expected_matrix[:, :5], decimal=6)
    np.testing.assert_array_equal(feature_names, np.array(['a', 'is', 'language', 'programming', 'python']))
```

File: tests/test_text_processing_util_mds24.py [hunk 5]
```diff
@@ -80,8 +83,84 @@ def test_list_num_punctuation():
...
+def test_tfidf_vectorizer_similar_content():
+    docs = ["machine learning is interesting", "machine learning is fascinating"]
+    tfidf_matrix, feature_names = tfidf_vectorizer(docs)
 
+    expected_matrix = np.array([
+        [ 0.,  0., -0.10136628, -0.10136628, -0.10136628],  
+        [ 0.,  0., -0.10136628, -0.10136628, -0.10136628] 
+    ])
+    assert_array_almost_equal(tfidf_matrix, expected_matrix[:, :5], decimal=6)
+    assert_array_equal(feature_names, np.array(['fascinating', 'interesting', 'is', 'learning', 'machine']))
```
Comment (jy1909) [inline tests/test_text_processing_util_mds24.py]: Add `np.testing.` because the extra import was removed.
```suggestion
    np.testing.assert_array_almost_equal(tfidf_matrix, expected_matrix[:, :5], decimal=6)
    np.testing.assert_array_equal(feature_names, np.array(['fascinating', 'interesting', 'is', 'learning', 'machine']))
```

File: tests/test_text_processing_util_mds24.py [hunk 6]
```diff
@@ -80,8 +83,84 @@ def test_list_num_punctuation():
...
+def test_tfidf_vectorizer_repeated_words():
+    docs = ["apple orange banana", "apple banana banana"]
+    tfidf_matrix, feature_names = tfidf_vectorizer(docs)
+
+    expected_matrix = np.array([
+        [-0.13515504, -0.13515504,  0.],  
+        [-0.13515504, -0.27031007,  0.] 
+    ])
+    assert_array_almost_equal(tfidf_matrix, expected_matrix, decimal=6)
+    assert_array_equal(feature_names, np.array(['apple', 'banana', 'orange']))
```
Comment (jy1909) [inline tests/test_text_processing_util_mds24.py]: Add `np.testing.` because the extra import was removed.
```suggestion
    np.testing.assert_array_almost_equal(tfidf_matrix, expected_matrix, decimal=6)
    np.testing.assert_array_equal(feature_names, np.array(['apple', 'banana', 'orange']))
```

File: tests/test_text_processing_util_mds24.py [hunk 7]
```diff
@@ -1,6 +1,9 @@
 import pytest
+import numpy as np
+from numpy.testing import assert_array_equal, assert_array_almost_equal
```
Comment (jy1909) [inline tests/test_text_processing_util_mds24.py]: Remove this. We can just add `np.testing.` before the assert functions. See below suggestions for examples.
```suggestion

```

**Review by jy1909 — COMMENTED** [review]
Great work calculating tf-idf! Just a few minor suggestions, and also fix the merge conflict.

**Review by jy1909 — APPROVED** [review]
